### ISSUE TITLE
[Security] There are still various ways to escape the eval limits

### ISSUE THREAD

**Issue by Alan-Liang** [AlexNisnevich/untrusted#465]
For example:

- [Overriding `onExit`](https://gist.github.com/Untrusted-Game/d6a6af8633edffd6e1e965bb2fccb7ce)
- [Escaping iframe](https://gist.github.com/Untrusted-Game/30a7a4b48dc6b197921c42c09093c9d7)

**Comment by pppery** [comment]
> [Overriding `onExit`](https://gist.github.com/Untrusted-Game/d6a6af8633edffd6e1e965bb2fccb7ce)
Should be fixed by #464 (which I wrote before seeing this issue, on general security grounds.)

**Comment by pppery** [comment]
FYI I personally think the bug you are exploiting in the "Overriding onExit" example is that you are allowed to close the `startLevel` function and run top level code at all, and not that you can then do nefarious things outside that function, but it's very unclear how much creativity of that sort should be allowed (this is a game focused on hacking the world after all) and how much should be forbidden.

**Comment by Alan-Liang** [comment]
> > [Overriding `onExit`](https://gist.github.com/Untrusted-Game/d6a6af8633edffd6e1e965bb2fccb7ce)
> > Should be fixed by #464 (which I wrote before seeing this issue, on general security grounds.)

Not really... it still works: <https://gist.github.com/Untrusted-Game/455bb325f49ddeb5c7dccda77f2709fd>

**Comment by pppery** [comment]
#464 hasn't been merged yet. I'm unable to reproduce the "Overriding onExit" trick if I apply that patch locally and test your code (note that it silently fails: the game doesn't produce a validation error of any sort, but the `onExit` override doesn't do anything and you still need to have the computer to exit the level.)

**Comment by Alan-Liang** [comment]
I regret for not noticing that.

于 2020年7月17日 GMT+08:00 下午11:45:18, pppery <notifications@github.com> 写到:
